consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/statemachine/QbftController.java

```diff
@@ -0,0 +1,108 @@
+import org.hyperledger.besu.consensus.common.bft.statemachine.BaseBlockHeightManager;
+import org.hyperledger.besu.consensus.common.bft.statemachine.BftFinalState;
+import org.hyperledger.besu.consensus.common.bft.statemachine.FutureMessageBuffer;
+import org.hyperledger.besu.consensus.qbft.messagedata.CommitMessageData;
+import org.hyperledger.besu.consensus.qbft.messagedata.PrepareMessageData;
+import org.hyperledger.besu.consensus.qbft.messagedata.ProposalMessageData;
+import org.hyperledger.besu.consensus.qbft.messagedata.QbftV1;
+import org.hyperledger.besu.consensus.qbft.messagedata.RoundChangeMessageData;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.Message;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+
+public class QbftController extends BaseBftController {
+
+  private BaseQbftBlockHeightManager currentHeightManager;
+  private final QbftBlockHeightManagerFactory qbftBlockHeightManagerFactory;
+
+  public QbftController(
+      final Blockchain blockchain,
+      final BftFinalState bftFinalState,
+      final QbftBlockHeightManagerFactory qbftBlockHeightManagerFactory,
+      final Gossiper gossiper,
+      final MessageTracker duplicateMessageTracker,
+      final FutureMessageBuffer futureMessageBuffer,
+      final SynchronizerUpdater sychronizerUpdater) {
+
+    super(
+        blockchain,
+        bftFinalState,
+        gossiper,
+        duplicateMessageTracker,
+        futureMessageBuffer,
+        sychronizerUpdater);
+    this.qbftBlockHeightManagerFactory = qbftBlockHeightManagerFactory;
+  }
+
+  @Override
+  protected void handleMessage(final Message message) {
+    final MessageData messageData = message.getData();
+
+    switch (messageData.getCode()) {
+      case QbftV1.PROPOSAL:
+        consumeMessage(
+            message,
+            ProposalMessageData.fromMessageData(messageData).decode(),
+            currentHeightManager::handleProposalPayload);
+        break;
+
+      case QbftV1.PREPARE:
+        consumeMessage(
+            message,
+            PrepareMessageData.fromMessageData(messageData).decode(),
+            currentHeightManager::handlePreparePayload);
+        break;
+
+      case QbftV1.COMMIT:
+        consumeMessage(
+            message,
+            CommitMessageData.fromMessageData(messageData).decode(),
+            currentHeightManager::handleCommitPayload);
+        break;
+
+      case QbftV1.ROUND_CHANGE:
+        consumeMessage(
+            message,
+            RoundChangeMessageData.fromMessageData(messageData).decode(),
+            currentHeightManager::handleRoundChangePayload);
+        break;
+
+      default:
+        throw new IllegalArgumentException(
+            String.format(
+                "Received message with messageCode=%d does not conform to any recognised IBFT message structure",
```

Comment:
change pare of log message from "IBFT message" to "QBFT message"

Reply:
done